### PR TITLE
usm: kafka: Support fetch v12

### PR DESCRIPTION
**Pull request by vitkyrka**
### What does this PR do?

Add support for v12 of the Fetch API.  The main change is the use of
varints and addition of tagged values.  While that change itself is
fairly straightforward, the need to parse varints caused the programs to
exceed the instruction size supported by older kernels (4096
instructions), so the programs had to be split into separate programs
for pre-v12 and v12, as well as to do the parsing in two steps in
separate programs: a first step to parse the locations of the record
batches arrays, and a second step to parse the list of found locations.

The split between pre-v12 and v12 is simple and handled by informing the 
compiler about the bounds of the versions, and then depending on it to
eliminate the unused branches.  The split of the parsing into two steps
requires code to handle the transition between the two steps correctly.

Also mainly due to code size restrictions on older kernels (even after
the above optimizations), the following limitations are currently
present:

 - A maximum of four bytes of varints are supported
 - The tagged values fields are always assumed to be empty

### Motivation

https://datadoghq.atlassian.net/browse/USMON-829

#### Maps change

`kafka_response` map increased from 6.92 MiB to 7.15 MiB (with 30k entries).

#### Load test results

Base version is main (`c4dd4ac`). Test type is `kafka.plain`.

system-probe max k8s CPU Usage  (3)
system-probe RSS (avg) (3)
process agent working set (avg)

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h [hunk 1]
```diff
@@ -255,14 +255,187 @@ static __always_inline enum parse_result read_with_remainder(kafka_response_cont
...
+    }
+}
+
+static __always_inline enum parse_result kafka_continue_parse_response_partition_loop(kafka_info_t *kafka,
                                                                             conn_tuple_t *tup,
                                                                             kafka_response_context_t *response,
                                                                             pktbuf_t pkt, u32 offset,
-                                                                            u32 data_end)
+                                                                            u32 data_end,
+                                                                            enum parser_level level,
```
Comment (DanielLavie) [inline pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:434]: I don't see `level` being used here

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h [hunk 2]
```diff
@@ -366,21 +571,105 @@ static __always_inline enum parse_result kafka_continue_parse_response_loop(kafk
...
+    // We should have exited at KAFKA_FETCH_RESPONSE_PARTITION_END if we
+    // managed to parse the entire packet, so if we get here we still have
+    // more to go. Remove the skb_info.data_off so that this function can
+    // be called again on the same packet with the same arguments in a tail
+    // call.
+    response->carry_over_offset = offset - orig_offset;
+    return RET_LOOP_END;
+}
+
+static __always_inline enum parse_result kafka_continue_parse_response_record_batches_loop(kafka_info_t *kafka,
```
Comment (DanielLavie) [inline pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:644]: I don't see `level` being used here

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h [hunk 3]
```diff
@@ -526,15 +839,150 @@ static __always_inline enum parse_result kafka_continue_parse_response_loop(kafk
     return RET_LOOP_END;
 }
 
-static __always_inline enum parse_result kafka_continue_parse_response(kafka_info_t *kafka,
+static __always_inline void kafka_call_response_parser(void *ctx, conn_tuple_t *tup, pktbuf_t pkt, kafka_response_state state, u32 api_version)
+{
+    enum parser_level level = parser_state_to_level(state);
+    u32 index;
```
Comment (DanielLavie) [inline pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:845]:
```suggestion
    u32 index = 0;
```

File: pkg/network/ebpf/c/protocols/kafka/types.h [hunk 4]
```diff
@@ -69,13 +81,24 @@ typedef struct kafka_response_context_t {
...
-    kafka_transaction_t transaction;
+    // The current value of the varint which we are processing.
+    __u32 varint_value;
+    // These could actually just be u8 (limited by
+    // KAFKA_MAX_RECORD_BATCHES_ARRAYS), but that leads to odd verifier errors
+    // when they are used in the handling of
+    // KAFKA_FETCH_RESPONSE_RECORD_BATCHES_ARRAY_END.
+    __u32 record_batches_arrays_idx;
+    __u32 record_batches_arrays_count;
+    // kafka_fetch_response_record_batches_array saved_record_batch_array;
```
Comment (DanielLavie) [inline pkg/network/ebpf/c/protocols/kafka/types.h:92]: Is this a leftover?

File: pkg/network/protocols/kafka/types.go [hunk 5]
```diff
@@ -21,6 +21,7 @@ type ConnTuple C.conn_tuple_t
 
 type EbpfTx C.kafka_event_t
 
+type KafkaFetchResponseRecordBatchesArray C.kafka_fetch_response_record_batches_array_t
```
Comment (DanielLavie) [inline pkg/network/protocols/kafka/types.go:24]: I don't see a usage of this type

**Review by DanielLavie — CHANGES_REQUESTED** [review]

File: pkg/network/usm/kafka_monitor_test.go [hunk 6]
```diff
@@ -168,6 +190,9 @@ func (s *KafkaProtocolParsingSuite) testKafkaProtocolParsing(t *testing.T, tls b
 		return getDefaultTestConfiguration(tls)
 	}
 
+	tmp, _ := version.LookupMaxKeyVersion(kafka.FetchAPIKey)
```
Comment (DanielLavie) [inline pkg/network/usm/kafka_monitor_test.go:193]: Could you incorporate handling for the scenario where the function returns false?

**Review by DanielLavie — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h (hunk 3 above)
Comment (DanielLavie) [reply pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:955]: I see, can you please add a comment in the code clarifying this point?

**Review by DanielLavie — APPROVED** [review]
Left a single nit comment

File: pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h (hunk 3 above)
Comment (vitkyrka) [reply pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:957]: Done.

**Pull request merged**